JSDK-2307 MSP objects do not recover on JSDK side after SFU Failover

makarandp0 opened the pull request:
Note: This PR was originally opened as https://github.com/twilio/twilio-video.js/pull/665, 

For network quality and dominant speaker updates client  listen on a datachannel. The id for the channels is obtained from RSP field (roomState.media_signaling.network_quality.transport.label, roomState.media_signaling.network_quality.transport.label).

During the call in a group room if the VMS process gets killed, a new instance gets created - This results in an update message on signaling channel with a new peerconnection id. Client responds by creating a new peer connection.

The problem was we were not rehooking our NetoworkMonitor, Dominant Spekear dataChannels from the new peerconnection in such cases.

With this change, we fix this by tearing down the monitors after close event on datachannel. They get recreated when a new update message is received.
I have verified the changes by killing the VMS and ensuring that we continue to update networkQualityLevelChanged, and Dominant Speaker changes.

- [x] I acknowledge that all my contributions will be made under the project's license.

manjeshbhargav commented on lib/signaling/v2/room.js:
```diff
@@ -340,6 +340,15 @@ class RoomV2 extends RoomSignaling {
         return;
       }
+
+      receiver.on('close', () => {
```
@makarandp0 How about:
```js
// NOTE(mpatwardhan): The underlying RTCDataChannel is closed whenever
// the VMS instance fails over, and a new RTCDataChannel is created in order
// to resume sending Dominant Speaker updates.
receiver.once('close', () => this._teardownDominantSpeakerSignaling());
```

manjeshbhargav commented on lib/signaling/v2/room.js:
```diff
@@ -363,6 +372,15 @@ class RoomV2 extends RoomSignaling {
         return;
       }
+
+      receiver.on('close', () => {
```
@makarandp0 How about:
```js
// NOTE(mpatwardhan): The underlying RTCDataChannel is closed whenever
// the VMS instance fails over, and new a RTCDataChannel is created in order
// to resume exchanging Network Quality messages.
receiver.once('close', () => this. _teardownNetworkQualityMonitor());
```

manjeshbhargav commented on lib/data/receiver.js:
```diff
@@ -37,6 +37,10 @@ class DataTrackReceiver extends DataTrackTransceiver {
       this.emit('message', event.data);
     });
+
+    dataChannel.addEventListener('close', () => {
```
@makarandp0 Can you add:
```js
/**
 * @emits DataTrackReceiver#close
 */
```
to the class documentation? Also, at the end of the file, can you define the "close" event:
```js
/**
 * @event DataTrackReceiver#close
 */
```

manjeshbhargav replied:
@makarandp0 Can you add a CHANGELOG.md entry?

manjeshbhargav commented on test/unit/spec/signaling/v2/room.js:
```diff
@@ -1335,6 +1335,48 @@ describe('RoomV2', () => {
           assert(participant.setNetworkQualityLevel.calledWith(0));
         });
 
+        describe('when the dataTrackReceiver receives close event', () => {
```
@makarandp0 
```js
describe('when the underlying DataTrackReceiver is closed', () => {
  ...
  it('should tear down the NetworkQualitySignaling', () => {
    ...
  });

  context('when a new DataTrackReceiver is created', () => {
    ...
    it(...);
    it(...);
    it(...);
  });
});
```

manjeshbhargav replied:
@makarandp0 Can you also add a test for Dominant Speaker signaling?

manjeshbhargav commented on test/unit/spec/signaling/v2/room.js:
```diff
@@ -1335,6 +1335,48 @@ describe('RoomV2', () => {
+            assert(NetworkQualitySignaling.calledWith(dataTrackReceiver2));
+          });
+
+          it('calls .start() on the NetworkQualityMonitor, and', () => {
```
@makarandp0 Nitpick: ~~, and,~~?

makarandp0 commented:
@manjeshbhargav , Thank you for looking at this. I have addressed your feedback, and added some tests for Dominant Speaker Handling. Can you take another look?

manjeshbhargav commented on lib/data/receiver.js:
```diff
@@ -55,6 +60,7 @@ class DataTrackReceiver extends DataTrackTransceiver {
 
 /**
  * @event DataTrackReceiver#message
+ * @event DataTrackReceiver#close
```
@makarandp0 "close" should be a separate block and should not be a part of this block.

manjeshbhargav commented on lib/signaling/v2/room.js:
```diff
@@ -340,7 +344,12 @@ class RoomV2 extends RoomSignaling {
+      // NOTE(mpatwardhan): The underlying RTCDataChannel is closed whenever
+      // the VMS instance fails over, and a new RTCDataChannel is created in order
+      // to resume sending Dominant Speaker updates.
+      receiver.on('close', () => this._teardownDominantSpeakerSignaling());
```
@makarandp0 "receiver.**once**(" ? Also can you add an extra line after this line. Usually its good to separate the code which has comments for it from the rest of the code.

manjeshbhargav commented on test/unit/spec/signaling/v2/room.js:
```diff
@@ -1184,6 +1185,118 @@ describe('RoomV2', () => {
     });
   });
 
+  // Dominant Speaker Signaling
```
@makarandp0 Remove comment lines.

manjeshbhargav commented on test/unit/spec/signaling/v2/room.js:
```diff
@@ -1184,6 +1185,118 @@ describe('RoomV2', () => {
+    });
+  });
+
   // Network Quality Signaling
```
@makarandp0 Remove comment lines.

manjeshbhargav commented on CHANGELOG.md:
```diff
@@ -24,7 +24,8 @@ New Features
 Bug Fixes
 ---------
-
+- Fixed a bug where, Dominant Speaker and Network Quality events used to stop
```
@makarandp0 
```
- Fixed a bug where Participants in a Group or Small Group Room stopped receiving Dominant Speaker
  and Network Quality updates when the media server recovered from a failover. (JSDK-2307)
```
